tokio/tests/time_delay_queue.rs

```diff
@@ -410,6 +410,91 @@ async fn insert_before_first_after_poll() {
     assert_eq!(entry, "two");
 }
 
+#[tokio::test]
+async fn reset_later_after_slot_starts() {
+    time::pause();
+
+    let mut queue = task::spawn(DelayQueue::new());
+
+    let now = Instant::now();
+
+    let foo = queue.insert_at("foo", now + ms(100));
+
+    assert_pending!(poll!(queue));
+
+    delay_for(ms(80)).await;
+
+    assert!(!queue.is_woken());
+
+    queue.reset_at(&foo, now + ms(120));
+    assert!(queue.is_woken());
```

Comment:
Should the task be woken here? Changing it from 100 to 120 while the current elapsed is at 80 doesn't seem like it should wake the task.

Reply:
It looks like because of the chosen times for this test, the normalized `when` ends up being less than `elapsed` which is why this wakeup happens. This is not actually the case.. if the new time is set to something like `200` instead of `120` this assert fails. I would therefore just remove this assertion.

Reply:
That assert is important to this test as the bug I'm trying to fix is that the task isn't woken at the appropriate time (or at all).  The timings are specifically chosen such that the new time is within a slot where the start time in the past; a test where the time is `200` would have passed without my changes.

In this case,
- at `t=0` the test adds an entry to fire at `t=100`, 
  - this gets added to the `[64-128)` slot
  - the queue creates a delay to fire at `t=100`
- at `t=80` the test resets the time on the entry to fire at `t=120`
  - the new entry also goes in the `[64-128)` slot
  - the queue resets the time on the delay to fire at the start of the first non-empty slot, which is `t=64` and hence in the past, so should fire immediately.

I'm keen to have the tests more closely represent the sequence of calls that would happen with an actual scheduler, that is (after the initial poll)
- only call `poll` if the task has been woken,
- call `poll` reasonably promptly after the task has been woken.

Both this bug and https://github.com/tokio-rs/tokio/pull/2217 were obscured by the fact that the tests call `poll` to check the state of the futures without checking that the task has been woken first.

I'll add some comments to explain why the particular numbers in these tests have been chosen, and why the task gets woken at that point.

Reply:
@thomaswhiteway I can review more of the timer internals today, but it's this note that I don't understand right now:

> the queue resets the time on the delay to fire at the start of the first non-empty slot, which is t=64 and hence in the past, so should fire immediately

If a timer level is 64 slots, then the `t=100` entry should be at slot 36 of the that level, and `t=120` should be at slot 56 of that level. Why would the slot 64 not be empty, and the `t=120` be added there?

Reply:
Never mind my confusion above I've continued stepping through the internals and this is making more sense.

I was not aware that the slots are stacks themselves, so it would make sense that the first slot holds 64 values--and why both `t=100` and `t=120` end up in the same slot.

I have a few more things to step through but don't worry about trying to response to the above comment!

Reply:
@thomaswhiteway Okay I'm a lot more familiar with the time internals now and see why this change is necessary.

There are a few subtle things here that helped in my understanding. I'm writing them out in case they are helpful for anyone else reviewing it that may not completely familiar with the internals:
* The delay queue initially sets it's `delay.expiration` to `t=101`. The corresponding `Entry` for this is in the `[64-128)` slot, but there is no attempt to normalize the value because we know the exact value.
* The delay queue and the timer driver have separate `Wheel`s that share `Entry`s. Therefore, even the the driver drives the initial `t=101` delay through it's `Wheel`, the delay queue's `delay.expiration` remains at `t=101`
* The `DelayQueue::reset_at` resets `delay.expiration` to a new value. Because we only know the first *slot* that is occupied, we set the new value to `t=65`.
* When `Entry::reset` is eventually called from the above, `when=65` (the new expiration value), `elapsed=81` (the driver has been updating this), and `curr=101` (the driver only updates this on on `fn fire` which has not happened yet).

In the state of that final bullet point, we are now setting the new expiration value to `t=65` (our best guess due to the first occupied slot) at `t=81` which should indicate this delay *may* be expired. Therefore, we need to wake the task that has registered interest in the delay.

All to say.. this looks good to me! Once you address the comment above about the loop, I can gives this a 👍.

Reply:
@kleimkuhler I've made the change to the loop.  I've also added some comments to the tests explaining the what's happening, as it's easy to miss the significance of the particular numbers chosen.

Reply:
Thanks!